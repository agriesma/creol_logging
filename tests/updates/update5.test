==========================================
reduce in PROGRAM : classes .
result Configuration: add(
< class("Log",0) : Class | Version: 0, Inh: noInh, Param: noVid, Att: "log" |->
    null, Mtds:
  < "init" : Method | Param: noVid, Att: "_" |-> null, "_" |-> null, "_" |->
    null, "_" |-> null, "_" |-> null, Code: skip ; return(emp) >, < "log" :
    Method | Param: "message", Att: "_" |-> null, "_" |-> null, "_" |-> null,
    "_" |-> null, "_" |-> null, "message" |-> null, "message" |-> null,
    "message" |-> null, "message" |-> null, "message" |-> null, Code: assign(
    "log" ; "|-|"("log" :: "message")) ; return(emp) >, < "run" : Method |
    Param: noVid, Att: "_" |-> null, "_" |-> null, "_" |-> null, "_" |-> null,
    "_" |-> null, Code: skip ; return(emp) >,
  Ocnt: 0 >
, none)
 extend("BankAccount", noInh, noSubst, < "deposit" : Method | Param: "sum",
    Att: "_" |-> null, "_" |-> null, "_" |-> null, "ccglab0" |-> null, "sum"
    |-> null, "sum" |-> null, "sum" |-> null, Code: call("ccglab0" ; "this" ;
    "transaction" ; "sum" :: "|-|"(str("deposit amount") :: "string_of_int"(
    "sum"))) ; return(emp) >, < "transfer" : Method | Param: "sum", "acc", Att:
    "_" |-> null, "_" |-> null, "_" |-> null, "acc" |-> null, "acc" |-> null,
    "acc" |-> null, "ccglab0" |-> null, "ccglab1" |-> null, "sum" |-> null,
    "sum" |-> null, "sum" |-> null, Code: await "<="("sum" :: "bal") ; call(
    "ccglab0" ; "this" ; "transaction" ; "-"("sum") :: "|-|"(str(
    "transfer amount") :: "string_of_int"("sum"))) ; get("ccglab0" ; noVid) ;
    call("ccglab1" ; "acc" ; "deposit" ; "sum") ; get("ccglab1" ; noVid) ;
    return(emp) >, < "withdraw" : Method | Param: "sum", Att: "_" |-> null, "_"
    |-> null, "_" |-> null, "ccglab0" |-> null, "sum" |-> null, "sum" |-> null,
    "sum" |-> null, Code: await "<="("sum" :: "bal") ; call("ccglab0" ; "this"
    ; "transaction" ; "-"("sum") :: "|-|"(str("withdraw amount") ::
    "string_of_int"("sum"))) ; get("ccglab0" ; noVid) ; return(emp) >, skip, c(
    "BankAccount",1))
 extend("BankAccount", noInh, "logger" |-> null, < "transaction" : Method |
    Param: "sum", "s", Att: "_" |-> null, "_" |-> null, "_" |-> null, "_" |->
    null, "ccglab0" |-> null, "s" |-> null, "s" |-> null, "s" |-> null, "s" |->
    null, "sum" |-> null, "sum" |-> null, "sum" |-> null, "sum" |-> null, Code:
    assign("bal" ; "+"("bal" :: "sum")) ; new("logger" ; "Log" ; emp) ; call(
    "ccglab0" ; "logger" ; "log" ; "s") ; get("ccglab0" ; noVid) ; return(emp)
    >, skip, c("BankAccount",0))
 remove("BankAccount", noInh, noSubst, < "increaseBalance" : Method | Param:
    "sum", Att: "_" |-> null, "_" |-> null, "sum" |-> null, "sum" |-> null,
    Code: skip >, c("BankAccount",2), none)
 
< class("BankAccount",0) : Class | Version: 8, Inh: noInh, Param: noVid, Att:
    "bal" |-> null, Mtds:
  < "balance" : Method | Param: noVid, Att: "_" |-> null, "_" |-> null, "_" |->
    null, "_" |-> null, "_" |-> null, "_" |-> null, "b" |-> null, Code: assign(
    "b" ; "bal") ; return("b") >, < "deposit" : Method | Param: "sum", Att: "_"
    |-> null, "_" |-> null, "_" |-> null, "_" |-> null, "_" |-> null, "_" |->
    null, "ccglab0" |-> null, "sum" |-> null, "sum" |-> null, "sum" |-> null,
    "sum" |-> null, "sum" |-> null, "sum" |-> null, Code: call("ccglab0" ;
    "this" ; "increaseBalance" ; "sum") ; free("ccglab0") ; return(emp) >, <
    "increaseBalance" : Method | Param: "sum", Att: "_" |-> null, "_" |-> null,
    "_" |-> null, "_" |-> null, "_" |-> null, "_" |-> null, "sum" |-> null,
    "sum" |-> null, "sum" |-> null, "sum" |-> null, "sum" |-> null, "sum" |->
    null, Code: assign("bal" ; "+"("bal" :: "sum")) ; return(emp) >, < "init" :
    Method | Param: noVid, Att: "_" |-> null, "_" |-> null, "_" |-> null, "_"
    |-> null, "_" |-> null, "_" |-> null, Code: assign("bal" @ "BankAccount" ;
    int(0)) ; return(emp) >, < "run" : Method | Param: noVid, Att: "_" |->
    null, "_" |-> null, "_" |-> null, "_" |-> null, "_" |-> null, "_" |-> null,
    Code: skip ; return(emp) >,
  Ocnt: 0 >

Bye.
