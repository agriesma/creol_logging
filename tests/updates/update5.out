==========================================
frewrite in PROGRAM : state .
result (sort not calculated): {
< label(ob("BankAccount-0"), 11) : Future |  Completed: true,  References: 1, 
    Value: emp >
 
< label(ob("BankAccount-0"), 13) : Future |  Completed: true,  References: 1, 
    Value: emp >
 
< label(ob("Customer-0"), 94) : Future |  Completed: false,  References: 0, 
    Value: emp >
 
< label(ob("Customer-0"), 98) : Future |  Completed: false,  References: 0, 
    Value: emp >
 
< label(ob("Customer-0"), 99) : Future |  Completed: false,  References: 0, 
    Value: emp >
 
< ob("BankAccount-0") : class("BankAccount",3) |
  Att: "bal" |-> int(100), "logger" |-> ob("Log-0"), "this" |-> ob(
    "BankAccount-0"),
  Pr: idle,
  PrQ: noProc,
  Lcnt: 14 >
 
< ob("Customer-0") : class("Customer",0) |
  Att: "account" |-> ob("BankAccount-0"), "this" |-> ob("Customer-0"), "timer"
    |-> int(5),
  Pr: {".class" |-> str("Customer"), ".label" |-> label(ob("Customer-0"), 98),
    ".method" |-> str("tick"), ".stage" |-> int(0), "_" |-> null, "caller" |->
    ob("Customer-0"), "ccglab0" |-> null | return(emp)},
  PrQ: {".class" |-> str("Customer"), ".label" |-> label(ob("Customer-0"), 99),
    ".method" |-> str("tick"), ".stage" |-> int(0), "_" |-> null, "caller" |->
    ob("Customer-0"), "ccglab0" |-> null | await "<"("timer" :: int(6)) ;
    assign("timer" ; "+"("timer" :: int(1))) ; call("ccglab0" ; "this" ; "tick"
    ; emp) ; free("ccglab0") ; return(emp)}, {".class" |-> str("Customer"),
    ".label" |-> label(ob("Customer-0"), 94), ".method" |-> str("trade"),
    ".stage" |-> int(0), "_" |-> null, "balance" |-> null, "caller" |-> ob(
    "Customer-0"), "ccglab0" |-> null, "ccglab1" |-> null, "ccglab2" |-> null |
    await ">"("timer" :: int(5)) ; assign("timer" ; int(0)) ; call("ccglab0" ;
    "account" ; "balance" ; emp) ; get("ccglab0" ; "balance") ; free("ccglab0")
    ; if "<="("balance" :: int(100)) th call("ccglab1" ; "account" ; "deposit"
    ; int(10)) ; get("ccglab1" ; noVid) ; free("ccglab1") el skip fi ; call(
    "ccglab2" ; "this" ; "trade" ; emp) ; free("ccglab2") ; return(emp)},
  Lcnt: 100 >
 
< ob("Log-0") : class("Log",1) |
  Att: "log" |-> list(str("deposit amount10")), "this" |-> ob("Log-0"),
  Pr: idle,
  PrQ: noProc,
  Lcnt: 2 >
 
< class("BankAccount",3) : Class | Version: 3, Inh: noInh, Param: noVid, Att:
    "bal" |-> null, "logger" |-> null, Mtds:
  < ".update" : Method | Param: noVid, Att: ".version" |-> int(0), "_" |->
    null, Code: skip >, < ".update" : Method | Param: noVid, Att: ".version"
    |-> int(1), "_" |-> null, Code: skip >, < "balance" : Method | Param:
    noVid, Att: "_" |-> null, "b" |-> null, Code: assign("b" ; "bal") ; return(
    "b") >, < "deposit" : Method | Param: "sum", Att: "_" |-> null, "ccglab0"
    |-> null, "sum" |-> null, Code: call("ccglab0" ; "this" ; "transaction" ;
    "sum" :: "|-|"(str("deposit amount") :: "string_of_int"("sum"))) ; return(
    emp) >, < "init" : Method | Param: noVid, Att: "_" |-> null, Code: assign(
    "bal" @ "BankAccount" ; int(0)) ; return(emp) >, < "run" : Method | Param:
    noVid, Att: "_" |-> null, Code: skip ; return(emp) >, < "transaction" :
    Method | Param: "sum", "s", Att: "_" |-> null, "ccglab0" |-> null, "s" |->
    null, "sum" |-> null, Code: assign("bal" ; "+"("bal" :: "sum")) ; if "="(
    "logger" :: null) th new("logger" ; "Log" ; emp) el skip fi ; call(
    "ccglab0" ; "logger" ; "log" ; "s") ; get("ccglab0" ; noVid) ; return(emp)
    >, < "transfer" : Method | Param: "sum", "acc", Att: "_" |-> null, "acc"
    |-> null, "ccglab0" |-> null, "ccglab1" |-> null, "sum" |-> null, Code:
    await "<="("sum" :: "bal") ; call("ccglab0" ; "this" ; "transaction" ; "-"(
    "sum") :: "|-|"(str("transfer amount") :: "string_of_int"("sum"))) ; get(
    "ccglab0" ; noVid) ; call("ccglab1" ; "acc" ; "deposit" ; "sum") ; get(
    "ccglab1" ; noVid) ; return(emp) >, < "withdraw" : Method | Param: "sum",
    Att: "_" |-> null, "ccglab0" |-> null, "sum" |-> null, Code: await "<="(
    "sum" :: "bal") ; call("ccglab0" ; "this" ; "transaction" ; "-"("sum") ::
    "|-|"(str("withdraw amount") :: "string_of_int"("sum"))) ; get("ccglab0" ;
    noVid) ; return(emp) >,
  Ocnt: 1 >
 
< class("Customer",0) : Class | Version: 0, Inh: noInh, Param: noVid, Att:
    "account" |-> null, "timer" |-> null, Mtds:
  < "init" : Method | Param: noVid, Att: "_" |-> null, Code: new("account" @
    "Customer" ; "BankAccount" ; emp) ; assign("timer" @ "Customer" ; int(0)) ;
    return(emp) >, < "run" : Method | Param: noVid, Att: "_" |-> null,
    "ccglab0" |-> null, "ccglab1" |-> null, Code: call("ccglab0" ; "this" ;
    "tick" ; emp) ; free("ccglab0") ; call("ccglab1" ; "this" ; "trade" ; emp)
    ; free("ccglab1") ; return(emp) >, < "tick" : Method | Param: noVid, Att:
    "_" |-> null, "ccglab0" |-> null, Code: await "<"("timer" :: int(6)) ;
    assign("timer" ; "+"("timer" :: int(1))) ; call("ccglab0" ; "this" ; "tick"
    ; emp) ; free("ccglab0") ; return(emp) >, < "trade" : Method | Param:
    noVid, Att: "_" |-> null, "balance" |-> null, "ccglab0" |-> null, "ccglab1"
    |-> null, "ccglab2" |-> null, Code: await ">"("timer" :: int(5)) ; assign(
    "timer" ; int(0)) ; call("ccglab0" ; "account" ; "balance" ; emp) ; get(
    "ccglab0" ; "balance") ; free("ccglab0") ; if "<="("balance" :: int(100))
    th call("ccglab1" ; "account" ; "deposit" ; int(10)) ; get("ccglab1" ;
    noVid) ; free("ccglab1") el skip fi ; call("ccglab2" ; "this" ; "trade" ;
    emp) ; free("ccglab2") ; return(emp) >,
  Ocnt: 1 >
 
< class("Log",1) : Class | Version: 1, Inh: noInh, Param: noVid, Att: "log" |->
    null, Mtds:
  < "init" : Method | Param: noVid, Att: "_" |-> null, Code: assign("log" @
    "Log" ; list(emp)) ; return(emp) >, < "log" : Method | Param: "message",
    Att: "_" |-> null, "message" |-> null, Code: assign("log" ; "|-"("log" ::
    "message")) ; return(emp) >, < "run" : Method | Param: noVid, Att: "_" |->
    null, Code: skip ; return(emp) >,
  Ocnt: 1 >
}
Bye.
